### PR TITLE
KDBX version 4 support and more

### PR DESCRIPTION
**Pull request by AlessioDP**
More info [here](https://github.com/AlessioDP/gokeepasslib).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,55 +1,20 @@
-gokeepasslib
+gokeepasslib (unstable)
```
Comment (tobischo) [inline README.md]: I assume it is stable now.

Do you want to keep these changes to the Pull Request?

Comment (AlessioDP) [reply README.md]: Sure why not, just make a new README like the original one :)

Comment (tobischo) [reply README.md]: Very well

**Comment by tobischo** [comment]
I apologize for not getting around to taking a look at the pull request sooner.

I created a new Database file using Keepass2 and checked how things would work by trying to integrate this state for gokeepasslib into kp2

#### Encryption: AES/Rijndael Key transformation: AES-KDF: 60000
- everything is fine

#### ChaCha20 Key transformation: AES-KDF: 60000
```
panic: cipher.NewCBCDecrypter: IV length must equal block size

goroutine 1 [running]:
crypto/cipher.NewCBCDecrypter(0x12a25a0, 0xc000082a80, 0xc0000883f0, 0xc, 0xc, 0x0, 0x0)
  /usr/local/go/src/crypto/cipher/cbc.go:107 +0x138
github.com/tobischo/kp2/vendor/github.com/tobischo/gokeepasslib.(*Database).Decrypter(0xc000082990, 0xc0000a8000, 0x20, 0x20, 0xe00, 0x0, 0x0, 0x1275028)
  /Users/tobiasschoknecht/go/src/github.com/tobischo/kp2/vendor/github.com/tobischo/gokeepasslib/database.go:57 +0xe0
github.com/tobischo/kp2/vendor/github.com/tobischo/gokeepasslib.decodeRawContent(0xc000082990, 0xc0002fb000, 0x824, 0xe00, 0xc0000a8000, 0x20, 0x20, 0x0, 0x0)
  /Users/tobiasschoknecht/go/src/github.com/tobischo/kp2/vendor/github.com/tobischo/gokeepasslib/decoder.go:99 +0x169
github.com/tobischo/kp2/vendor/github.com/tobischo/gokeepasslib.(*Decoder).Decode(0xc0000c3c20, 0xc000082990, 0xc000096060, 0x0)
  /Users/tobiasschoknecht/go/src/github.com/tobischo/kp2/vendor/github.com/tobischo/gokeepasslib/decoder.go:58 +0x15b
main.loadDatabaseCmd(0xc0000e0000, 0xc00009c2e0, 0x1, 0x2, 0x0, 0x0)
  /Users/tobiasschoknecht/go/src/github.com/tobischo/kp2/prerun.go:47 +0x254
github.com/tobischo/kp2/vendor/github.com/spf13/cobra.(*Command).execute(0xc0000e0000, 0xc00009c280, 0x2, 0x2, 0xc0000e0000, 0xc00009c280)
  /Users/tobiasschoknecht/go/src/github.com/tobischo/kp2/vendor/github.com/spf13/cobra/command.go:751 +0x521
github.com/tobischo/kp2/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc0000e0c80, 0x8, 0x4, 0xc0000e0a00)
  /Users/tobiasschoknecht/go/src/github.com/tobischo/kp2/vendor/github.com/spf13/cobra/command.go:852 +0x2fd
github.com/tobischo/kp2/vendor/github.com/spf13/cobra.(*Command).Execute(0xc0000e0c80, 0xc0000c3f60, 0x5)
  /Users/tobiasschoknecht/go/src/github.com/tobischo/kp2/vendor/github.com/spf13/cobra/command.go:800 +0x2b
main.main()
  /Users/tobiasschoknecht/go/src/github.com/tobischo/kp2/main.go:128 +0x543
```
which makes sense since we currently only initialize an AES cipher:
```go
// database.go#L51
// Decrypter initializes a CBC decrypter for the database
func (db *Database) Decrypter(transformedKey []byte) (cipher.BlockMode, error) {
	block, err := aes.NewCipher(buildMasterKey(db, transformedKey))
	if err != nil {
		return nil, err
	}
	return cipher.NewCBCDecrypter(block, db.Header.FileHeaders.EncryptionIV), nil
}
```

#### Encryption: AES/Rijndael Key transformation: Argon2 Iterations: 2, Memory: 1MB, Parellelism: 2
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x11dd580]

goroutine 1 [running]:
github.com/tobischo/kp2/vendor/github.com/tobischo/gokeepasslib.(*Decoder).Decode(0xc0000c3c20, 0xc000084990, 0xc000098060, 0x0)
	/Users/tobiasschoknecht/go/src/github.com/tobischo/kp2/vendor/github.com/tobischo/gokeepasslib/decoder.go:44 +0x3b0
main.loadDatabaseCmd(0xc0000e0000, 0xc00009e2e0, 0x1, 0x2, 0x0, 0x0)
	/Users/tobiasschoknecht/go/src/github.com/tobischo/kp2/prerun.go:47 +0x254
github.com/tobischo/kp2/vendor/github.com/spf13/cobra.(*Command).execute(0xc0000e0000, 0xc00009e280, 0x2, 0x2, 0xc0000e0000, 0xc00009e280)
	/Users/tobiasschoknecht/go/src/github.com/tobischo/kp2/vendor/github.com/spf13/cobra/command.go:751 +0x521
github.com/tobischo/kp2/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc0000e0c80, 0x8, 0x4, 0xc0000e0a00)
	/Users/tobiasschoknecht/go/src/github.com/tobischo/kp2/vendor/github.com/spf13/cobra/command.go:852 +0x2fd
github.com/tobischo/kp2/vendor/github.com/spf13/cobra.(*Command).Execute(0xc0000e0c80, 0xc0000c3f60, 0x5)
	/Users/tobiasschoknecht/go/src/github.com/tobischo/kp2/vendor/github.com/spf13/cobra/command.go:800 +0x2b
main.main()
	/Users/tobiasschoknecht/go/src/github.com/tobischo/kp2/main.go:128 +0x543
```

[Database.kdbx.zip](https://github.com/tobischo/gokeepasslib/files/2633933/Database.kdbx.zip)
Password: test123test

I am not sure when I will have time to look into this some more. Just leaving this here in case you want to take a closer look

**Comment by AlessioDP** [comment]
Now it should works, I've implemented ChaCha20 encryption/decryption.
All tests are okay, I've added ChaCha20 test too 👍

**Comment by tobischo** [comment]
I found the issue I had. It was caused by my implementation in kp2.
Up until your changes it wasn't necessary to have options initialized so `new(gokeepasslib.Database)` worked. That does not work anymore for v4 files 😄 

I added a test for argon2 decoding

**Pull request merged**